test/Test/Database/LSMTree/Internal/Integration.hs

```diff
@@ -0,0 +1,199 @@
+import           Data.Coerce (coerce)
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe
+import           Data.Primitive.ByteArray (ByteArray (ByteArray),
+                     sizeofByteArray)
+import qualified Data.Set as Set
+import qualified Data.Vector.Primitive as P
+import           Data.Word
+import           Database.LSMTree.Generators
+import           Database.LSMTree.Internal.BlobRef (BlobSpan)
+import           Database.LSMTree.Internal.Entry
+import           Database.LSMTree.Internal.Integration
+import           Database.LSMTree.Internal.RawPage
+import           Database.LSMTree.Internal.Run.BloomFilter as Bloom
+import           Database.LSMTree.Internal.Run.Construction as Run
+import           Database.LSMTree.Internal.Run.Index.Compact as Index
+import           Database.LSMTree.Internal.Serialise
+import           Database.LSMTree.Internal.Serialise.Class
+import           Database.LSMTree.Internal.Serialise.RawBytes as RB
+import           Database.LSMTree.Util
+import           GHC.Generics
+import           Test.Database.LSMTree.Generators (deepseqInvariant,
+                     prop_arbitraryAndShrinkPreserveInvariant)
+import           Test.QuickCheck
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+import           Test.Util.Orphans ()
+import           Test.Util.QuickCheck as Util.QC
+
+tests :: TestTree
+tests = testGroup "Test.Database.LSMTree.Internal.Integration" [
+      testGroup "InMemLookupData" $ prop_arbitraryAndShrinkPreserveInvariant (deepseqInvariant @InMemLookupData)
+    , localOption (QuickCheckMaxSize 1000) $
+      testProperty "prop_inMemRunLookupAndConstruction" $ prop_inMemRunLookupAndConstruction
+    ]
+
+-- | Construct a run incrementally, then test a number of positive and negative lookups.
+prop_inMemRunLookupAndConstruction :: InMemLookupData -> Property
+prop_inMemRunLookupAndConstruction dat =
+    Map.size runData > 0 ==>
+        tabulateKeySizes
+      $ tabulateValueSizes
+      $ tabulateNumKeyEntryPairs
+      $ tabulateNumPages
+      $ conjoin (fmap check prepped) .&&. conjoin (fmap check' prepped')
+  where
+    InMemLookupData{runData, lookups} = dat
+
+    tabulateKeySizes = tabulate "Size of key in run" [showPowersOf10 $ sizeofKey k | k <- Map.keys runData ]
+    tabulateValueSizes = tabulate "Size of value in run" [showPowersOf10 $ onValue 0 sizeofValue e | e <- Map.elems runData]
+    tabulateNumKeyEntryPairs = tabulate "Number of key-entry pairs" [showPowersOf10 (Map.size runData) ]
+    tabulateNumPages = tabulate "Number of pages" [showPowersOf10 (Map.size ps) | let (ps, _, _) = run]
+
+    run = mkTestRun runData
+    -- prepLookups says that a key /could be/ in the given page
+    prepped = prepLookups [run] lookups
+    -- prepLookups says that a key /is definitely not/ in the given page
+    prepped' = Set.toList (Set.fromList lookups Set.\\ Set.fromList (fmap fst prepped))
```

Comment:
These names could be a bit clearer, maybe something like `keysToLookup` and `keysNotToLookup`? And something similar for `check` and `check'`.